pkg/controller/sparkapplication/controller.go

```diff
@@ -500,21 +532,33 @@ func (c *Controller) processSingleExecutorStateUpdate(update *executorStateUpdat
 
 func (c *Controller) updateSparkApplicationStatusWithRetries(
 	original *v1alpha1.SparkApplication,
-	updateFunc func(*v1alpha1.SparkApplicationStatus)) *v1alpha1.SparkApplication {
-	toUpdate := original.DeepCopy()
+	updateFunc func(*v1alpha1.SparkApplicationStatus),
+) *v1alpha1.SparkApplication {
 
 	var lastUpdateErr error
 	for i := 0; i < maximumUpdateRetries; i++ {
-		updated, err := c.tryUpdateStatus(original, toUpdate, updateFunc)
+		toUpdate := original.DeepCopy()
+
+		glog.V(2).Infof("Trying to update SparkApplication %s", original.Name)
+		// Apply update
+		updateFunc(&toUpdate.Status)
+
+		// Let's keep the old App status if this is not a valid transition.
+		if !isValidDriverStatusTransition(original.Status.AppState.State, toUpdate.Status.AppState.State) {
+			toUpdate.Status.AppState = original.Status.AppState
+		}
+		updated, err := c.tryUpdateStatus(original, toUpdate)
 		if err == nil {
+			c.postMetrics(original, updated)
 			return updated
 		}
 		lastUpdateErr = err
+		glog.Errorf("[Attempt: %v] failed to update SparkApplication [%v] %v", i, toUpdate, err)
```

Comment:
The first `%v` can be `%d`. Also do you really want to log the entire `SparkApplication` object instead of just the app name?

Reply:
I wanted to log more details since this is an error state. For now, I changed this to just log the name. Will probably add more details once we have the state management change out and this should be pretty rare